beamer/check/commands.py

```diff
@@ -0,0 +1,256 @@
+)
+@click.option(
+    "--rpc-file",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to the RPC config file.",
+)
+@click.option(
+    "--artifacts-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="The directory containing deployment artifacts.",
+)
+@click.option(
+    "--abi-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="Path to the directory with contract ABIs.",
+)
+@click.option(
+    "--output",
+    type=click.Path(file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to store the invalidation info at, which can be later used for verification.",
+)
+@click.option(
+    "--count",
+    type=click.IntRange(1),
+    default=1,
+    show_default=True,
+    help="Number of invalidations to create, per (PROOF_SOURCE, PROOF_TARGET) pair.",
+)
+@click.argument("proof-source", type=ChainIdParam())
+@click.argument("proof-target", type=ChainIdParam(), nargs=-1, required=True)
+def initiate_l1_invalidations(
+    keystore_file: Path,
+    password: str,
+    rpc_file: Path,
+    artifacts_dir: Path,
+    abi_dir: Path,
+    output: Path,
+    count: int,
+    proof_source: ChainId,
+    proof_target: tuple[ChainId],
+) -> None:
+    """Initiate L1 invalidations from chain PROOF_SOURCE to each PROOF_TARGET.
+    PROOF_SOURCE and PROOF_TARGET are chain IDs."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    account = beamer.util.account_from_keyfile(keystore_file, password)
+    log.info("Loaded keystore file", address=account.address)
+
+    abi_manager = ABIManager(abi_dir)
+    deployment = beamer.artifacts.load(artifacts_dir, proof_source)
+
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+    url = rpc_info[proof_source]
+    w3 = make_web3(url, account)
+    assert w3.eth.chain_id == proof_source
+    log.info("Connected to RPC", url=url)
+
+    fill_manager = obtain_contract(w3, abi_manager, deployment, "FillManager")
+    request_manager = obtain_contract(w3, abi_manager, deployment, "RequestManager")
+    finality_period = request_manager.functions.chains(proof_source).call()[0]
+
+    # Step 1: issue invalidateFill transactions.
+    # This will produce invalidations without finalization_timestamp.
+    #
+    # Note that if the output file already contains invalidations from
+    # proof_source to a specific proof_target, we will only issue  as many new
+    # invalidations as is necessary to reach `count`.
+    if output.exists():
+        with output.open("rt") as f:
+            data = json.load(f)
+        invalidations = apischema.deserialize(list[Invalidation], data)
+    else:
+        invalidations = []
+
+    existing_counts: Counter[ChainId] = Counter()
+    for invalidation in invalidations:
+        if invalidation.proof_source == proof_source:
```

Comment:
Question: 

Do you intent to batch multiple proof sources into one global end file or do you plan to leave the generated outputs (invalidation files), which are separated by proof_source.

If one or the other it would be ideal to either keep the if condition here, or to actually assert if the consumed output file is compatible.

Reply:
> Do you intent to batch multiple proof sources into one global end file or do you plan to leave the generated outputs (invalidation files), which are separated by proof_source.

The intent was/is to allow for invalidations with different proof sources in the same file, i.e. I would not prohibit having different proof sources in the same file. This lets the user decide the policy with respect to how those files are organized, named etc. Most likely, for the toplevel script, we'd just have 1 file per proof_source to all other chains, but I wanted to leave that up to the script, that is, the user of the command.

> If one or the other it would be ideal to either keep the if condition here,

Yeah, I'd like for this condition to stay there.

Your question made me realize that for a differing proof_source we will do nothing in step 1, while in step 2 we may actually run the relayer, regardless of both proof_source and proof_target. So I'm thinking step 2 about skipping step 2 too if proof_source of the invalidation does not match and proof_target of the invalidation is not in the provided list. What's your opinion on this?

Reply:
> Your question made me realize that for a differing proof_source we will do nothing in step 1, while in step 2 we may actually run the relayer, regardless of both proof_source and proof_target. So I'm thinking step 2 about skipping step 2 too if proof_source of the invalidation does not match and proof_target of the invalidation is not in the provided list. What's your opinion on this?

That's a good point.. I see two arguments for the two sides. If we automatically do prove calls, we always require all RPCs in the rpc file, even though the command was about different chains. This could be a problem of not being able to successfully finish a command cause of an unrelated missing rpc.
On the other hand, it's easy to miss finalizations. 

My suggestion would be to clearly indicate that invalidations for proof_source X have been completed, to give at least some feedback to the user in this direction, but only handle those invalidations which are part of the command

Reply:
> My suggestion would be to clearly indicate that invalidations for proof_source X have been completed, to give at least some feedback to the user in this direction, but only handle those invalidations which are part of the command

Yeah, that sounds good. Will add something to the logs and adjust the current behaviour to only consider specified proof source. :+1:

Reply:
Here's an example of (1) running with all invalidations already sent, and (2) running with 1 invalidation sent, the rest already done before (makes for 4 in total).

```
beamer-py3.10) [ivan@beta4 beamer]$ beamer check initiate-l1-invalidations --keystore-file 0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc.json --password '' --rpc-file .local.rpc.json --artifacts-dir deployments/artifacts/test/ --abi-dir contracts/.build/ -
output /tmp/foo.invalidations 1337 1337 --count 3 
023-08-21 14:04:59.899602 [info     ] Loaded keystore file           [beamer.check.commands] address=0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc
023-08-21 14:04:59.966677 [info     ] Connected to RPC               [beamer.check.commands] url=http://127.0.0.1:8545
023-08-21 14:05:00.008475 [info     ] Invalidations initiated successfully [beamer.check.commands] path=/tmp/foo.invalidations sent=0 total=3
beamer-py3.10) [ivan@beta4 beamer]$ beamer check initiate-l1-invalidations --keystore-file 0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc.json --password '' --rpc-file .local.rpc.json --artifacts-dir deployments/artifacts/test/ --abi-dir contracts/.build/ -
output /tmp/foo.invalidations 1337 1337 --count 4 
023-08-21 14:05:14.263204 [info     ] Loaded keystore file           [beamer.check.commands] address=0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc
023-08-21 14:05:14.319749 [info     ] Connected to RPC               [beamer.check.commands] url=http://127.0.0.1:8545
023-08-21 14:05:14.422818 [info     ] Transaction sent               [beamer.check.commands] block=85 txhash=0x040cca501a506d62aa6922ddae1bd0e8901a724b6fb165463916c08504a77145
023-08-21 14:05:14.423418 [info     ] Sent invalidation              [beamer.check.commands] idx=3 proof_source=1337 proof_target=1337
023-08-21 14:05:14.441107 [info     ] Invalidation complete          [beamer.check.commands] finalization_timestamp=2023-08-21T14:05:14 idx=3 proof_source=1337 proof_target=1337
023-08-21 14:05:14.441248 [info     ] Invalidations initiated successfully [beamer.check.commands] path=/tmp/foo.invalidations sent=1 total=4
```

Reply:
looks good!